[realsense2] Update to v2.19.0

claudiofantacci opened the pull request:
This PR updates `librealsense` to version `2.19.0`.

It is still WIP and proper check should be made, especially because it partially reverts #4564, and in particular this https://github.com/Microsoft/vcpkg/pull/4564/commits/6afbceb27fd2aa39c66f810a22f5d4ad97575110.

Under windows, the `debug/bin` files were erroneously deleted and the installation failed.

Closes #5753.

claudiofantacci committed:
[realsense2] Update to v2.19.0

vicroms commented:
Regression Status (commit https://github.com/Microsoft/vcpkg/pull/5777/commits/4e6531a10c6c980aec0b33d5087b340f24854bc4)

platform|port|master| test |notes
--|--|--|--|--
x64-windows|realsense2 |Pass  |Fail  |**Regression**
x64-windows-static|realsense2        |Pass  |Fail  |**Regression**
x86-windows|realsense2 |Pass  |Fail  |**Regression**


x64-osx had no regressions
arm64-windows had no regressions
x64-linux had no regressions
x64-uwp had no regressions
arm-uwp had no regressions

claudiofantacci commented:
Hi @vicroms, I don't understand the table 😄 
Master passed, but test failed and I'm not allowed (error 401) to check `vcpkg-PR-Eager`.
Can I have an hint to improve the PR?

As a side note, I'd like to change `-DCMAKE_DEBUG_POSTFIX=_d` in `-DCMAKE_DEBUG_POSTFIX=d` to be consistent with the majority of libraries present in vcpkg when compiled in debug mode.

Rastaban commented:
@claudiofantacci, The port is building in the master branch, but fails in the tests run against this PR.  Here is the relevent log for x64-windows:

```
Starting package 1/1: realsense2:x64-windows
Building package realsense2[core]:x64-windows... 
Could not locate cached archive: C:\vsts\_work\2\s\archives\db\db31abea05c8de6a9d6877d3f52e25a8e71dfb3d.zip
-- Downloading https://github.com/IntelRealSense/librealsense/archive/v2.19.0.tar.gz...
-- Extracting source C:/vsts/_work/2/s/downloads/IntelRealSense-librealsense-v2.19.0.tar.gz
-- Using source at C:/vsts/_work/2/s/buildtrees/realsense2/src/v2.19.0-3a2934e71a
-- Configuring x64-windows
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:56 (message):
    Command failed: ninja;-v
    Working Directory: C:/vsts/_work/2/s/buildtrees/realsense2/x64-windows-rel/vcpkg-parallel-configure
    Error code: 1
    See logs for more information:
      C:\vsts\_work\2\s\buildtrees\realsense2\config-x64-windows-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_cmake.cmake:269 (vcpkg_execute_required_process)
  ports/realsense2/portfile.cmake:21 (vcpkg_configure_cmake)
  scripts/ports.cmake:71 (include)


Error: Building package realsense2:x64-windows failed with: BUILD_FAILED
Elapsed time for package realsense2:x64-windows: 23.98 s
```

[config-x64-windows-out.log](https://github.com/Microsoft/vcpkg/files/3005446/config-x64-windows-out.log)

If you are interested in seeing any other logs don't hesitate to ask.

vicroms commented:
It is the same problem as PR #5275. 
`realsense2` tries to obtain `libusb` using Git instead of using `libusb` installed by VCPKG.

claudiofantacci commented:
@vicroms yes, it is one of the problem I'm trying to understand. The issue involves `ExternalProject_Add` calls in `librealsense`'s `CMake/external_libusb.cmake`. It works perfectly when used with VS, but fails using `ninja` build for some strange permission errors. I found a way to let it work with `ninja`, but it requires a different folder configuration inside the build tree directory. Anyway, why are the tests run with `ninja` build even if the port is not configured to do so?

claudiofantacci commented:
I force pushed an update to this PR.
As described in #5753, I added a patch, removed `-DBUILD_WITH_TM2=OFF` and we now use `ninja`.
Let me know test reports 👍 

Note that I also changed `-DCMAKE_DEBUG_POSTFIX=_d` to `-DCMAKE_DEBUG_POSTFIX=d` to agree with all the orther libraries. Is it ok?

claudiofantacci commented:
It seems that tests are still failing 😞
@Rastaban may I please have a look at the logs?

Rastaban commented:
Sure! the regressions are still on x64-windows, x64-windows-static, and x86-windows.
[failureLogs.zip](https://github.com/Microsoft/vcpkg/files/3010108/failureLogs.zip)

claudiofantacci commented:
Errors:
 1. **x64-windows**: _ExternalProject_ - error: could not find git for clone of libusb.
 2. **x64-windows-static**: _ExternalProject_ - error: could not find git for clone of libusb.
 3. **x86-windows**: _ExternalProject_ - error: could not find git for clone of libusb.

The builds are not able to get `libusb`, which is required to build the library, form the interent using git.
It is not related to either the library or a CMake error but, I guess, on the build environment for the tests. Are you aware of any possible constraints related to this?

What I can do is to disable a specific option during CMake configuration and apply another patch to the library. I'm discussing this upstream.

claudiofantacci commented:
Upstream discussion about `librealsense` issues with Windows: https://github.com/IntelRealSense/librealsense/issues/3597

claudiofantacci commented:
@Rastaban and @vicroms, do you have any feedback for me about my last post https://github.com/Microsoft/vcpkg/pull/5777#issuecomment-477025200?

claudiofantacci commented:
In the meantime that I'm discussing upstream for Windows fixes, I pushed a new commit with another patch to see if this works (at least temporarily).

ras0219-msft commented on ports/realsense2/portfile.cmake:
```diff
@@ -14,48 +17,47 @@ string(COMPARE EQUAL "${VCPKG_CRT_LINKAGE}" "static" BUILD_CRT_LINKAGE)
+      # BUILD
+      -DBUILD_EXAMPLES=${BUILD_EXAMPLES}
+      -DBUILD_GRAPHICAL_EXAMPLES=${BUILD_GRAPHICAL_EXAMPLES}
+      -DBUILD_SHARED_LIBS=${BUILD_LIBRARY_LINKAGE}
```
This setting is automatically provided by vcpkg_configure_cmake and doesn't require special provisions in the portfile.

claudiofantacci replied:
Thanks for the feedback. I'm removing `BUILD_SHARED_LIBS` from the port.

ras0219-msft commented on ports/realsense2/portfile.cmake:
```diff
@@ -14,48 +17,47 @@ string(COMPARE EQUAL "${VCPKG_CRT_LINKAGE}" "static" BUILD_CRT_LINKAGE)
+      # Ungrouped Entries
+      -DENFORCE_METADATA=ON
+      # BUILD
+      -DBUILD_EXAMPLES=${BUILD_EXAMPLES}
```
This should be moved to `OPTIONS_RELEASE` since it may otherwise conflict with the below `OPTIONS_DEBUG`

claudiofantacci replied:
I totally agree. I just left it there as it was in the current version of the port. Moved 👍

vicroms committed:
Merge branch 'master' into upd/librealsense

claudiofantacci commented:
A brief update on this: all the problems related to the Windows build have been fixed upstream.
As a consequence, with the next stable release (`> 2.21.0`) the portfile should be a cleaner, stable and with Ninja build system. For a detailed view of the many fixes applied upstream, you can start from https://github.com/IntelRealSense/librealsense/pull/3611.

vicroms commented:
@claudiofantacci 
Thanks for the update!

vicroms commented:
A pre-release tag for v2.22.0 has been created 🎉

claudiofantacci commented:
Great! I think we can update the port as soon as we have a stable release 👍

vicroms committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into upd/librealsense

vicroms committed:
[realsense2] Update to version 2.22.0

vicroms commented:
Updated port to version 2.22.0

Waiting for CI results.

claudiofantacci commented:
> Updated port to version 2.22.0
> 
> Waiting for CI results.

Great!

Please note that upgrading to 2.22.0 the following option no longer exists:
```cmake
-DBUILD_WITH_TM2=OFF
```

You can probably check from the CMake log in the CI 👍

vicroms committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into upd/librealsense

vicroms commented:
@claudiofantacci  Thanks for your awesome work!
